doc/srcfiles.1
.\" Copyright 2023 Red Hat Inc.
.\" Mon 2023-Sept 23 Housam Alamour <halamour@redhat.com>
.\" Contact elfutils-devel@sourceware.org to correct errors or typos.
.TH EU-SRCFILES 1 "2023-Sept-25" "elfutils"

.de SAMPLE
.br
.RS 0
.nf
.nh
\fB
..
.de ESAMPLE
\fP
.hy
.fi
.RE
..

.SH "NAME"
eu-srcfiles \- Lists the source files of a DWARF/ELF file.

.SH "SYNOPSIS"
eu-srcfiles [\fB\-0\fR|\fB\-\-null\fR] [\fB\-c\fR|\fB\-\-cu\-only\fR] [\fB\-v\fR|\fB\-\-verbose\fR] INPUT

.SH "DESCRIPTION"
\fBeu-srcfiles\fR lists the source files of a given \s-DWARF/ELF\s0
file.  This list is based on a search of the DWARF debuginfo, which
may be automatically fetched by debuginfod if applicable.  The target
file may be an executable, a coredump, a process, or even the running
kernel.  The default is the file 'a.out'.  The source file names are
made unique and printed to standard output.

.SH "INPUT OPTIONS"
The long and short forms of options, shown here as alternatives, are
equivalent.
.TP
\fB--core=COREFILE\fR
Find addresses from signatures found in COREFILE.

.TP
\fB--debuginfo-path=PATH\fR
Search path for separate debuginfo files.

.TP
\fB-e FILE\fR, \fB--executable=FILE\fR
Find addresses in FILE.

.TP
\fB-k\fR, \fB--kernel\fR
Find addresses in the running kernel.

.TP
\fB-K\fR, \fB--offline-kernel[=RELEASE]\fR
Kernel with all modules.

.TP
\fB-M FILE\fR, \fB--linux-process-map=FILE\fR
Find addresses in files mapped as read from FILE in Linux /proc/PID/maps format.

.TP
\fB-p PID\fR, \fB--pid=PID\fR
Find addresses in files mapped into process PID.

.TP
\fB-?\fR, \fB--help\fR
Give this help list.

.TP
\fB--usage\fR
Give a short usage message.

.TP
\fB-V\fR, \fB--version\fR
Print program version.

.SH "OUTPUT OPTIONS"

.TP
\fB\-0, \-\-null\fR
Separate items by a null instead of a newline.

.TP
\fB\-c, \-\-cu\-only\fR
Only list the CU names.

.TP
\fB\-v, \-\-verbose\fR
Increase verbosity of logging messages.


.SH EXAMPLES

List all source files for a binary.
.SAMPLE
eu-srcfiles -e /bin/ls
.ESAMPLE

List all compilation units (CU) names for a given process (including shared libraries).
.SAMPLE
eu-srcfiles -c -p $$
.ESAMPLE

List source files of a binary based on its buildid, using debuginfod.
.SAMPLE
binary=`debuginfod-find executable 9c22d8d9e42bd051ffdc1064fdfd456ba781c629`
eu-srcfiles -c -e $binary
.ESAMPLE

Show the source code of the first CU of a shared library.
.SAMPLE
binary=/usr/lib64/libc.so.6
srcfile=`eu-srcfiles -c -e $binary | head -1`
cat `debuginfod-find source $binary $srcfile`
.ESAMPLE

List the source files of a kernel image.
.SAMPLE
eu-srcfiles -e /boot/vmlinuz-`uname -r`
.ESAMPLE


.SH "AUTHOR"
Written by Housam Alamour.

.SH "REPORTING BUGS"
Please reports bugs at https://sourceware.org/bugzilla/

.SH "COPYRIGHT"
Copyright (c) 2023 Red Hat Inc.  License GPLv3+: GNU GPL version 3 or
later <https://gnu.org/licenses/gpl.html>.  This is free software: you
are free to change and redistribute it.  There is NO WARRANTY, to the
extent permitted by law.


NEWS
Version 0.191 (after 0.190)

debuginfod: Caching eviction logic improvements to improve retention
            of small/frequent/slow files such as Fedora's vdso.debug.

srcfiles: Can now fetch the source files of a DWARF/ELF file and
          place them into a zip.

Version 0.190 "Woke!"

CONTRIBUTING: Switch from real name policy to known identity policy.
              Updated ChangeLog policy (no more separate ChangeLog files).
              There is a SECURITY bug policy now.
              The default branch is now 'main'.

libelf: Add RELR support.

libdw: Recognize .debug_[ct]u_index sections

readelf: Support readelf -Ds, --use-dynamic --symbol.
         Support .gdb_index version 9

scrlines: New tool that compiles a list of source files associated
          with a specified dwarf/elf file.

debuginfod: Schema change (reindexing required, sorry!) for a 60%
            compression in filename representation, which was a large
            part of the sqlite index; also, more deliberate sqlite
            -wal management during scanning using the
            --scan-checkpoint option.

backends: Various LoongArch updates.

Version 0.189 "Don't deflate!"

configure: eu-nm, eu-addr2line and eu-stack can provide demangled symbols
           when linked with libstdc++. Use --disable-demangler to disable.

           A new option --enable-sanitize-memory has been added for msan
           sanitizer support.

libelf: elf_compress now supports ELFCOMPRESS_ZSTD when build against
        libzstd

libdwfl: dwfl_module_return_value_location now returns 0 (no return type)
         for DIEs that point to a DW_TAG_unspecified_type.

elfcompress: -t, --type= now support zstd if libelf has been build with
             ELFCOMPRESS_ZSTD support.

backends: Add support for LoongArch and Synopsys ARCv2 processors.

Version 0.188 "no section left behind"

readelf: Add -D, --use-dynamic option.

debuginfod-client: Add $DEBUGINFOD_HEADERS_FILE setting to supply outgoing
                   HTTP headers. Add new function debuginfod_find_section.

debuginfod: Add --disable-source-scan option.

libdwfl: Add new function dwfl_get_debuginfod_client.
         Add new function dwfl_frame_reg.
         Add new function dwfl_report_offline_memory.

Version 0.187 "no zero negatives"

debuginfod: Support -C option for connection thread pooling.

debuginfod-client: Negative cache file are now zero sized instead of
                   no-permission files.

addr2line: The -A, --absolute option, which shows file names including
           the full compilation directory is now the default.  To get the
           old behavior use the new option --relative.

readelf, elflint: Recognize FDO Packaging Metadata ELF notes

libdw, debuginfo-client: Load libcurl lazily only when files need to
                         be fetched remotely. libcurl is now never
                         loaded when DEBUGINFOD_URLS is unset. And when
                         DEBUGINFOD_URLS is set, libcurl is only loaded
                         when the debuginfod_begin function is called.

Version 0.186 "Just In Time"

debuginfod-client: Default $DEBUGINFOD_URLS is computed from drop-in files
                   /etc/debuginfod/*.urls rather than hardcoded into the
                   /etc/profile.d/debuginfod* scripts.
                   Add $DEBUGINFOD_MAXSIZE and $DEBUGINFOD_MAXTIME settings
                   for skipping large/slow transfers.
                   Add $DEBUGINFOD_RETRY for retrying aborted lookups.

debuginfod: Supply extra HTTP response headers, describing archive/file
            names that satisfy the requested buildid content.
            Support -d :memory: option for in-memory databases.
            Protect against loops in federated server configurations.
            Add -r option to use -I/-X regexes for grooming stale files.
            Protect against wasted CPU from duplicate concurrent requests.
            Limit the duration of groom ops roughly to rescan (-t) times.
            Add --passive mode for serving from read-only database.
            Several other performance improvements & prometheus metrics.

libdw: Support for the NVIDIA Cuda line map extensions.
       DW_LNE_NVIDIA_inlined_call and DW_LNE_NVIDIA_set_function_name
       are defined in dwarf.h. New functions dwarf_linecontext and
       dwarf_linefunctionname

translations: Update Japanese translation.

Version 0.185 "oops, I did it again"

debuginfod-client: Simplify curl handle reuse so downloads which
                   return an error are retried.

elfcompress: Always exit with code 0 when the operation succeeds (even
             when nothing was done). On error the exit code is now always 1.

Version 0.184 "negative results"

debuginfod: Use libarchive's bsdtar as the .deb-family file unpacker.

debuginfod-client: Client caches negative results. If a query for a
		   file failed with 404, an empty 000 permission
		   file is created in the cache. This will prevent
		   requesting the same file for the next 10 minutes.

		   Client objects now carry long-lived curl handles
		   for outgoing connections.  This makes it more
		   efficient for multiple sequential queries, because
		   the TCP connections and/or TLS state info are kept
		   around awhile, avoiding O(100ms) setup latencies.

libdw: handle DW_FORM_indirect when reading attributes

translations: Update Polish translation.

Version 0.183 "100+ commits"

debuginfod: New thread-busy metric and more detailed error metrics.
            New --fdcache-mintmp and tracking of filesystem freespace.
            New increased webapi concurrency while grooming.

debuginfod-client: DEBUGINFOD_SONAME macro added to debuginfod.h which
                   can be used to dlopen the libdebuginfod.so library.
                   New function debuginfod_set_verbose_fd and
                   DEBUGINFOD_VERBOSE environment variable.

config: profile.sh and profile.csh won't export DEBUGINFOD_URLS unless

src/srcfiles.cxx
/* Print the source files of a given ELF file.
   Copyright (C) 2023 Red Hat, Inc.
   This file is part of elfutils.
   Written by Housam Alamour <alamourh@redhat.com>.

   This file is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   elfutils is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "printversion.h"
#include <dwarf.h>
#include <argp.h>
#include <cstring>
#include <set>
#include <string>
#include <cassert>
#include <config.h>

#include <libdwfl.h>
#include <fcntl.h>
#include <iostream>
#include <libdw.h>

using namespace std;

/* Name and version of program.  */
ARGP_PROGRAM_VERSION_HOOK_DEF = print_version;

/* Bug report address.  */
ARGP_PROGRAM_BUG_ADDRESS_DEF = PACKAGE_BUGREPORT;

/* Definitions of arguments for argp functions.  */
static const struct argp_option options[] =
{
  { NULL, 0, NULL, OPTION_DOC, N_("Output options:"), 1 },
  { "null", '0', NULL, 0,
    N_ ("Separate items by a null instead of a newline."), 0 },
  { "verbose", 'v', NULL, 0,
    N_ ("Increase verbosity of logging messages."), 0 },
  { "cu-only", 'c', NULL, 0, N_ ("Only list the CU names."), 0 },
  { NULL, 0, NULL, 0, NULL, 0 }
};

/* Short description of program.  */
static const char doc[] = N_("Lists the source files of a DWARF/ELF file. The default input is the file 'a.out'.");

/* Strings for arguments in help texts.  */
static const char args_doc[] = N_("INPUT");

/* Prototype for option handler.  */
static error_t parse_opt (int key, char *arg, struct argp_state *state);

static struct argp_child argp_children[2]; /* [0] is set in main.  */

/* Data structure to communicate with argp functions.  */
static const struct argp argp =
{
  options, parse_opt, args_doc, doc, argp_children, NULL, NULL
};

/* Verbose message printing. */
static bool verbose;
/* Delimit the output with nulls. */
static bool null_arg;
/* Only print compilation unit names. */
static bool CU_only;

/* Handle program arguments. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Suppress "unused parameter" warning. */
  (void)arg;
  switch (key)
    {
    case ARGP_KEY_INIT:
      state->child_inputs[0] = state->input;
      break;

    case '0':
      null_arg = true;
      break;

    case 'v':
      verbose = true;
      break;

    case 'c':
      CU_only = true;
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}


/* Global list of collected source files.  Normally, it'll contain
   the sources of just one named binary, but the '-K' option can cause
   multiple dwfl modules to be loaded, thus listed.   */
   set<string> debug_sourcefiles;

static int
collect_sourcefiles (Dwfl_Module *dwflmod,
                     void **userdata __attribute__ ((unused)),
                     const char *name __attribute__ ((unused)),
                     Dwarf_Addr base __attribute__ ((unused)),
                     void *arg __attribute__ ((unused)))
{
  Dwarf *dbg;
  Dwarf_Addr bias; /* ignored - for addressing purposes only  */

  dbg = dwfl_module_getdwarf (dwflmod, &bias);

  Dwarf_Off offset = 0;
  Dwarf_Off old_offset;
  size_t hsize;

  /* Traverse all CUs of this module.  */
  while (dwarf_nextcu (dbg, old_offset = offset, &offset, &hsize, NULL, NULL, NULL) == 0)
    {
      Dwarf_Die cudie_mem;
      Dwarf_Die *cudie = dwarf_offdie (dbg, old_offset + hsize, &cudie_mem);

      if (cudie == NULL)
        continue;

      const char *cuname = dwarf_diename (cudie) ?: "<unknown>";
      Dwarf_Files *files;
      size_t nfiles;
      if (dwarf_getsrcfiles (cudie, &files, &nfiles) != 0)
        continue;

      /* extract DW_AT_comp_dir to resolve relative file names  */
      const char *comp_dir = "";
      const char *const *dirs;
      size_t ndirs;

      if (dwarf_getsrcdirs (files, &dirs, &ndirs) == 0 && dirs[0] != NULL)
        comp_dir = dirs[0];
      if (comp_dir == NULL)
        comp_dir = "";

      if (verbose)
        std::clog << "searching for sources for cu=" << cuname
                  << " comp_dir=" << comp_dir << " #files=" << nfiles
                  << " #dirs=" << ndirs << endl;

      for (size_t f = 1; f < nfiles; f++)
        {
          const char *hat;
          if (CU_only)
          {
            if (strcmp(cuname, "<unknown>") == 0 || strcmp(cuname, "<artificial>") == 0 )
              continue;
            hat = cuname;
          }
          else
            hat = dwarf_filesrc (files, f, NULL, NULL);

          if (hat == NULL)
            continue;

          if (string(hat).find("<built-in>")
              != std::string::npos) /* gcc intrinsics, don't bother record  */
            continue;

          string waldo;
          if (hat[0] == '/') /* absolute */
            waldo = (string (hat));
          else if (comp_dir[0] != '\0') /* comp_dir relative */
            waldo = (string (comp_dir) + string ("/") + string (hat));
          debug_sourcefiles.insert (waldo);
        }
    }

  return DWARF_CB_OK;
}


int
main (int argc, char *argv[])
{
  int remaining;

  /* Parse and process arguments.  This includes opening the modules.  */
  argp_children[0].argp = dwfl_standard_argp ();
  argp_children[0].group = 1;

  Dwfl *dwfl = NULL;
  (void) argp_parse (&argp, argc, argv, 0, &remaining, &dwfl);
  assert (dwfl != NULL);
  /* Process all loaded modules - probably just one, except if -K or -p is used. */
  (void) dwfl_getmodules (dwfl, &collect_sourcefiles, NULL, 0);

  if (!debug_sourcefiles.empty ())
    for (const string &element : debug_sourcefiles)
      {
        std::cout << element;
        if (null_arg)
          std::cout << '\0';
        else
          std::cout << '\n';
      }

  dwfl_end (dwfl);
  return 0;
}
